js/server/modules/@arangodb/foxx/manager.js

```diff
@@ -660,57 +642,48 @@ function _deleteServiceFromPath (mount, options) {
 // FoxxService.basePath(mount) and FoxxService.bundlePath(mount). This is to
 // avoid a race condition with selfHeal() which could delete the files before
 // the service is saved to the database.
-function _install (mount, tempPaths, options = {}) {
+function _install (tempService, tempBundlePath, options = {}) {
   try {
-    const {tempServicePath: servicePath, tempBundlePath: bundlePath} = tempPaths;
-    const collection = utils.getStorage();
-    let service;
-    try {
-      service = FoxxService.create({
-        mount,
-        options,
-        basePath: servicePath,
-        noisy: true
-      });
-      if (options.setup !== false) {
-        service.executeScript('setup');
+    if (options.setup !== false) {
+      try {
+        tempService.executeScript('setup');
+      } catch (e) {
+        if (!options.force) {
+          e.codeFrame = codeFrame(e, tempService.basePath);
+          throw e;
+        } else {
+          console.warnStack(e);
+        }
       }
+    }
+    // instead of service.updateChecksum(), update the checksum
+    // manually from the temporary path.
+    tempService.checksum = FoxxService._checksumPath(tempBundlePath);
+    try {
+      ensureServiceExecuted(tempService, true);
     } catch (e) {
       if (!options.force) {
-        e.codeFrame = codeFrame(e, servicePath);
+        e.codeFrame = codeFrame(e, tempService.basePath);
         throw e;
       } else {
         console.warnStack(e);
       }
     }
-    // instead of service.updateChecksum(), update the checksum
-    // manually from the temporary path.
-    service.checksum = FoxxService._checksumPath(bundlePath);
     const bundleCollection = utils.getBundleStorage();
-    if (!bundleCollection.exists(service.checksum)) {
-      bundleCollection._binaryInsert({_key: service.checksum}, bundlePath);
+    if (!bundleCollection.exists(tempService.checksum)) {
+      bundleCollection._binaryInsert({_key: tempService.checksum}, tempBundlePath);
     }
-    const serviceDefinition = service.toJSON();
-    const meta = db._query(aql`
-      UPSERT {mount: ${mount}}
+    const serviceDefinition = tempService.toJSON();
+    const collection = utils.getStorage();
+    db._query(aql`
+      UPSERT {mount: ${tempService.mount}}
       INSERT ${serviceDefinition}
       REPLACE ${serviceDefinition}
       IN ${collection}
-      RETURN NEW
-    `).next();
-    service._rev = meta._rev;
-    GLOBAL_SERVICE_MAP.get(db._name()).set(mount, service);
```

Comment:
Storing the temporary service instance might save us a few millis on the first request but it's riskier to have a request potentially served by a service with a non-existent `basePath` than just let the `reloadRouting` take care later.